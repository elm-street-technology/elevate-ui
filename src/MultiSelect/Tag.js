import React from "react";
import classNames from "classnames";
import withStyles from "react-jss";

import Close from "elevate-ui-icons/Close";

type Item = {
  label: string,
  value: string,
};

type Props = {
  classes: Object,
  className: string,
  tag: Item,
  onRemove: Function,
};

const Tag = (props: Props) => {
  const { classes, className, tag, onRemove } = props;
  return (
    <div className={classNames(classes.root, className)}>
      <span className={classes.text}>{tag.label}</span>
      <button
        type="button"
        className={classes.button}
        onClick={(e) => {
          e.stopPropagation();
          e.preventDefault();
          onRemove(tag);
        }}
      >
        <Close size={18} />
      </button>
    </div>
  );
};

const styles = (theme) => ({
  root: {
    display: "flex",
    alignItems: "center",
    border: `1px solid ${theme.colors.secondary}`,
    borderRadius: theme.globalBorderRadius,
    margin: "2px",
  },
  text: {
    fontSize: "16px",
    lineHeight: "24px",
    fontWeight: "600",
    color: theme.colors.secondary,
    padding: "2px 2px 2px 8px",
  },
  button: {
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    color: theme.colors.secondary,
    padding: "2px",
  },
});

export default withStyles(styles, { name: "EuiTag" })(Tag);
