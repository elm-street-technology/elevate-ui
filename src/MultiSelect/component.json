{
  "description": "A component that renders a <MultiSelect /> to be used inside forms.",
  "displayName": "MultiSelect",
  "methods": [
    {
      "name": "stateReducer",
      "docblock": null,
      "modifiers": [],
      "params": [
        {
          "name": "state",
          "type": null
        },
        {
          "name": "changes",
          "type": null
        }
      ],
      "returns": null
    },
    {
      "name": "handleStateChange",
      "docblock": null,
      "modifiers": [],
      "params": [
        {
          "name": "changes",
          "type": null
        },
        {
          "name": "downshiftStateAndHelpers",
          "type": null
        }
      ],
      "returns": null
    },
    {
      "name": "onInputChange",
      "docblock": null,
      "modifiers": [],
      "params": [
        {
          "name": "e",
          "type": null
        }
      ],
      "returns": null
    },
    {
      "name": "onAddTag",
      "docblock": null,
      "modifiers": [],
      "params": [
        {
          "name": "item",
          "type": null
        }
      ],
      "returns": null
    },
    {
      "name": "onRemoveTag",
      "docblock": null,
      "modifiers": [],
      "params": [
        {
          "name": "item",
          "type": null
        }
      ],
      "returns": null
    },
    {
      "name": "onInputKeyDown",
      "docblock": null,
      "modifiers": [],
      "params": [
        {
          "name": "event",
          "type": null
        }
      ],
      "returns": null
    },
    {
      "name": "popValue",
      "docblock": null,
      "modifiers": [],
      "params": [],
      "returns": null
    },
    {
      "name": "onWrapperClick",
      "docblock": null,
      "modifiers": [],
      "params": [
        {
          "name": "e",
          "type": null
        }
      ],
      "returns": null
    },
    {
      "name": "focusOnInput",
      "docblock": null,
      "modifiers": [],
      "params": [],
      "returns": null
    },
    {
      "name": "inputRef",
      "docblock": null,
      "modifiers": [],
      "params": [
        {
          "name": "c",
          "type": null
        }
      ],
      "returns": null
    },
    {
      "name": "inputWrapperRef",
      "docblock": null,
      "modifiers": [],
      "params": [
        {
          "name": "c",
          "type": null
        }
      ],
      "returns": null
    },
    {
      "name": "renderArrowIcon",
      "docblock": null,
      "modifiers": [],
      "params": [
        {
          "name": "isOpen",
          "type": null
        }
      ],
      "returns": null
    }
  ],
  "props": {
    "classes": {
      "flowType": {
        "name": "Object"
      },
      "required": true,
      "description": ""
    },
    "className": {
      "flowType": {
        "name": "string"
      },
      "required": true,
      "description": "Accepts any classNames to be passed down to the component."
    },
    "closeOnSelect": {
      "flowType": {
        "name": "boolean"
      },
      "required": true,
      "description": "Whether or not the input should close when a value is selected."
    },
    "field": {
      "flowType": {
        "name": "Object"
      },
      "required": true,
      "description": ""
    },
    "form": {
      "flowType": {
        "name": "Object"
      },
      "required": true,
      "description": ""
    },
    "items": {
      "flowType": {
        "name": "Array",
        "elements": [
          {
            "name": "signature",
            "type": "object",
            "raw": "{\n  label: string,\n  value: string,\n}",
            "signature": {
              "properties": [
                {
                  "key": "label",
                  "value": {
                    "name": "string",
                    "required": true
                  }
                },
                {
                  "key": "value",
                  "value": {
                    "name": "string",
                    "required": true
                  }
                }
              ]
            }
          }
        ],
        "raw": "Array<Item>"
      },
      "required": true,
      "description": ""
    },
    "label": {
      "flowType": {
        "name": "string"
      },
      "required": true,
      "description": "Text input for the label used inside the component."
    },
    "onSelect": {
      "flowType": {
        "name": "Function"
      },
      "required": false,
      "description": "Accepts a function to be passed down to the component that fires when selected."
    },
    "onRemove": {
      "flowType": {
        "name": "Function"
      },
      "required": false,
      "description": "Accepts a function to be passed down to the component that fires when a tag is removed."
    },
    "tags": {
      "flowType": {
        "name": "boolean"
      },
      "required": false,
      "description": "Whether or not custom tags can be added to the MultiSelect."
    },
    "theme": {
      "flowType": {
        "name": "Object"
      },
      "required": true,
      "description": ""
    },
    "withScaffold": {
      "flowType": {
        "name": "boolean"
      },
      "required": true,
      "description": "If the scaffold should be used."
    }
  }
}