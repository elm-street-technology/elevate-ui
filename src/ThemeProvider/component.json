{
  "description": "",
  "displayName": "ThemeProvider",
  "methods": [],
  "props": {
    "withReset": {
      "defaultValue": {
        "value": "true",
        "computed": false
      },
      "required": false
    }
  }
}