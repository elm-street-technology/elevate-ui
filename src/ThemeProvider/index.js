// @flow
import React from "react";
import injectSheet, { ThemeProvider as JSSThemeProvider } from "react-jss";
import merge from "lodash/merge";

const colors = {
  primary: "#f15953",
  primaryDark: "#d24f45",
  primaryGlow: "#d24f45", // todo
  secondary: "#55c3ba",
  secondaryDark: "#44b0a7",
  tertiary: "#2e7fc2",
  quaternary: "#ffc629",
  white: "#fff",
  gray050: "#fbfcfd",
  gray100: "#edf1f4",
  gray200: "#dde2e7",
  gray300: "#ccd2d8",
  gray400: "#b3bac1",
  gray500: "#888f96",
  gray600: "#60686f",
  gray700: "#4b565f",
  gray800: "#5a5b5c",
  gray900: "#232c35",
  black: "#11181e",
  error: "#9c2929",
  facebook: "#3b5998",
  google: "#db4437",
  linkedin: "#007bb5",
  pinterest: "#bd081c",
  twitter: "#1da1f2",
};

const alertColors = {
  error: {
    color: "#9c2929",
    backgroundColor: "#fcf2f2",
    borderColor: "#f6dddd",
  },
  info: {
    color: "#0c5460",
    backgroundColor: "#d1ecf1",
    borderColor: "#bee5eb",
  },
  success: {
    color: "#246761",
    backgroundColor: "#ecf8f7",
    borderColor: "#d2e8e6",
  },
  warning: {
    color: "#856404",
    backgroundColor: "#fff3cd",
    borderColor: "#ffeeba",
  },
};

const typography = {
  bodyColor: "#2E2E35",
  fontFamily:
    "'Open Sans', -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif",
  fontSize: "1rem",
};

const transitions = {
  default: "all 200ms linear",
};

const breakpoints = (minWidth: number) => {
  if (!minWidth) {
    return new Error("Must enter a valid minWidth to use theme breakpoints");
  }
  return `@media (min-width: ${minWidth}px)`;
};

const globalBorderRadius = "4px";
const globalBoxShadow =
  "0 2px 5px 0 rgba(0,0,0,0.16), 0 2px 10px 0 rgba(0,0,0,0.12)";
const globalPadding = {
  padding: "12px",
  [breakpoints.tabletPortrait]: {
    padding: "16px",
  },
  [breakpoints.tabletLandscape]: {
    padding: "20px",
  },
  [breakpoints.desktop]: {
    padding: "24px",
  },
};

const shadowKeyUmbraOpacity = 0.2;
const shadowKeyPenumbraOpacity = 0.14;
const shadowAmbientShadowOpacity = 0.12;

function createShadow(...px) {
  return [
    `${px[0]}px ${px[1]}px ${px[2]}px ${
      px[3]
    }px rgba(0, 0, 0, ${shadowKeyUmbraOpacity})`,
    `${px[4]}px ${px[5]}px ${px[6]}px ${
      px[7]
    }px rgba(0, 0, 0, ${shadowKeyPenumbraOpacity})`,
    `${px[8]}px ${px[9]}px ${px[10]}px ${
      px[11]
    }px rgba(0, 0, 0, ${shadowAmbientShadowOpacity})`,
  ].join(",");
}

const shadows = [
  createShadow(0, 1, 3, 0, 0, 1, 1, 0, 0, 2, 1, -1),
  createShadow(0, 1, 5, 0, 0, 2, 2, 0, 0, 3, 1, -2),
  createShadow(0, 1, 8, 0, 0, 3, 4, 0, 0, 3, 3, -2),
  createShadow(0, 2, 4, -1, 0, 4, 5, 0, 0, 1, 10, 0),
  createShadow(0, 3, 5, -1, 0, 5, 8, 0, 0, 1, 14, 0),
  createShadow(0, 3, 5, -1, 0, 6, 10, 0, 0, 1, 18, 0),
  createShadow(0, 4, 5, -2, 0, 7, 10, 1, 0, 2, 16, 1),
  createShadow(0, 5, 5, -3, 0, 8, 10, 1, 0, 3, 14, 2),
  createShadow(0, 5, 6, -3, 0, 9, 12, 1, 0, 3, 16, 2),
  createShadow(0, 6, 6, -3, 0, 10, 14, 1, 0, 4, 18, 3),
  createShadow(0, 6, 7, -4, 0, 11, 15, 1, 0, 4, 20, 3),
  createShadow(0, 7, 8, -4, 0, 12, 17, 2, 0, 5, 22, 4),
  createShadow(0, 7, 8, -4, 0, 13, 19, 2, 0, 5, 24, 4),
  createShadow(0, 7, 9, -4, 0, 14, 21, 2, 0, 5, 26, 4),
  createShadow(0, 8, 9, -5, 0, 15, 22, 2, 0, 6, 28, 5),
  createShadow(0, 8, 10, -5, 0, 16, 24, 2, 0, 6, 30, 5),
  createShadow(0, 8, 11, -5, 0, 17, 26, 2, 0, 6, 32, 5),
  createShadow(0, 9, 11, -5, 0, 18, 28, 2, 0, 7, 34, 6),
  createShadow(0, 9, 12, -6, 0, 19, 29, 2, 0, 7, 36, 6),
  createShadow(0, 10, 13, -6, 0, 20, 31, 3, 0, 8, 38, 7),
  createShadow(0, 10, 13, -6, 0, 21, 33, 3, 0, 8, 40, 7),
  createShadow(0, 10, 14, -6, 0, 22, 35, 3, 0, 8, 42, 7),
  createShadow(0, 11, 14, -7, 0, 23, 36, 3, 0, 9, 44, 8),
  createShadow(0, 11, 15, -7, 0, 24, 38, 3, 0, 9, 46, 8),
];

const zIndex = {
  buttonBase: 0,
  buttonChildren: 1,
  selectDropdown: 2,
  tableLoading: -1,
  tableLoadingActive: 2,
  tableResizer: 10,
  tooltip: 1000,
  modal: 2000,
};

const defaultTheme = {
  alertColors,
  breakpoints,
  colors,
  globalBorderRadius,
  globalBoxShadow,
  globalPadding,
  shadows,
  transitions,
  typography,
  zIndex,
};

const GlobalsAndReset = injectSheet((theme) => ({
  "@global": {
    /* Eric Meyer Reset v2.0 */
    /*Â https://meyerweb.com/eric/tools/css/reset/ */
    "html, body, div, span, applet, object, iframe, h1, h2, h3, h4, h5, h6, p, blockquote, pre, a, abbr, acronym, address, big, cite, code, del, dfn, em, img, ins, kbd, q, s, samp, small, strike, strong, sub, sup, tt, var, b, u, i, center, dl, dt, dd, ol, ul, li, fieldset, form, label, legend, table, caption, tbody, tfoot, thead, tr, th, td, article, aside, canvas, details, embed, figure, figcaption, footer, header, hgroup, menu, nav, output, ruby, section, summary, time, mark, audio, video": {
      margin: "0",
      padding: "0",
      border: "0",
      fontSize: "100%",
      font: "inherit",
      verticalAlign: "baseline",
    },
    "article, aside, details, figcaption, figure, footer, header, hgroup, menu, nav, section": {
      display: "block",
    },
    body: {
      lineHeight: "1",
    },
    "ol, ul": {
      listStyle: "none",
    },
    "blockquote, q": {
      quotes: "none",
    },
    "blockquote:before, blockquote:after, q:before, q:after": {
      content: "''",
      content: "none", // eslint-disable-line
    },
    table: {
      borderCollapse: "collapse",
      borderSpacing: "0",
    },
    /* Globals */
    "html, body": {
      width: "100%",
      minHeight: "100%",
      fontFamily: theme.typography.fontFamily || "sans-serif",
      fontSize: theme.typography.fontSize || "1rem",
      fontWeight: "500",
      color: theme.typography.bodyColor || "#000",
      backgroundColor: "#fafafa",
      "-webkit-font-smoothing": "antialiased",
      "-moz-osx-font-smoothing": "grayscale",
      margin: "0",
      padding: "0",
    },
    /* Box-sizing border-box */
    /* https://www.paulirish.com/2012/box-sizing-border-box-ftw/ */
    html: {
      boxSizing: "border-box",
      background: "transparent",
    },
    "*, *:before, *:after": {
      boxSizing: "inherit",
    },
    /* Reset `button` to nothing */
    /* https://gist.github.com/chrisheninger/a860f87ef4e529b2df606768b97665a8 */
    button: {
      display: "inline-block",
      "-webkit-appearance": "none",
      "-moz-appearance": "none",
      userSelect: "none",
      textDecoration: "none",
      fontFamily: "inherit",
      fontSize: "inherit",
      color: "inherit",
      cursor: "pointer",
      border: "none",
      backgroundColor: "transparent",
      borderColor: "transparent",
      padding: "0",
      margin: "0",
      "&:disabled": {
        opacity: 0.6,
        cursor: "not-allowed",
      },
    },
  },
}))(({ children }) => children);

const ThemeProvider = ({ children, theme, withReset = true, ...rest }: any) => {
  const mergedTheme = theme ? merge(defaultTheme, theme) : defaultTheme;
  return (
    <JSSThemeProvider theme={mergedTheme} {...rest}>
      {withReset ? <GlobalsAndReset>{children}</GlobalsAndReset> : children}
    </JSSThemeProvider>
  );
};

export default ThemeProvider;
