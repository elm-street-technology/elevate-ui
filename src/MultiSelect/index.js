// @flow
import React, { Component, Fragment } from "react";
import withStyles from "../withStyles";
import classNames from "classnames";
import Downshift from "downshift";
import AutosizeInput from "react-input-autosize";
import get from "lodash/get";

import Scaffold from "../Scaffold";
import Tag from "./Tag";

type Item = {
  label: string,
  value: string,
};
type Items = Array<Item>;

type Props = {
  classes: Object,
  /**
   * Accepts any classNames to be passed down to the component.
   */
  className: string,
  /**
   * Whether or not the input should close when a value is selected.
   */
  closeOnSelect: boolean,
  field: Object,
  form: Object,

  /**
   * Sample usage:
   *
   * onSearch({inputValue, originalItems, setItems}) => {
   *   if (inputValue === '') {
   *     return setItems(originalItems);
   *   }
   *   return axios.get(`/your-autocomplete-endpoint?keywords=${inputValue}`)
   *    .then(({data: {newItems}}) => {
   *       return setItems(newItems); // replaces currentItems (if any exist)
   *    });
   * }
   *
   * render() {
   *    return <Multiselect onSearch={this.onSearch} items={this.state.multiselectItems} />;
   * }
   */
  onSearch?: ({
    inputValue: string,
    originalItems: Items,
    currentItems: Items,
    setItems: (Items) => void,
  }) => Items,

  /**
   * Accepts an array of objects with the shape { label: 'Banana', value: 'Banana' }
   */
  items: Items,
  /**
   * Text input for the label used inside the component.
   */
  label: string,
  /**
   * Whether or not custom tags can be added to the MultiSelect.
   * Tags are strings that the user can type and select that are not included in Items
   */
  tags?: boolean,
  theme: Object,
  /**
   * If the scaffold should be used.
   */
  withScaffold: boolean,
};

type State = {
  /**
   * Current input state.
   */
  inputValue: string,
  selectedItems: Items,

  // We move items into state here so that we can change or add items
  // Via autocomplete
  items: Items,
};

const itemToString = (item) => (item ? item.label : "");
const getSelectedItems = (props) => {
  const {
    field: { value },
    items,
  } = props;

  let selectedItems = [];

  // Sometimes, we might have some this.props.field.values that aren't
  // part of the this.props.items collection that gets passed in. We'll want
  // to add these so we have access to them in this.state.selectedItems
  // (they won't show up in the dropdown items but they'll show as selected)
  if (value && value.length > 0) {
    value.forEach((singleValue) => {
      const existingItem = items.find(
        (item) => singleValue.indexOf(item.value) > -1
      );
      if (existingItem) {
        selectedItems.push(existingItem);
      } else {
        selectedItems.push({
          label: singleValue,
          value: singleValue,
        });
      }
    });
  }

  return selectedItems;
};

/**
 * A component that renders a <MultiSelect /> to be used inside forms.
 */
class MultiSelect extends Component<Props, State> {
  constructor(props) {
    super(props);

    this.state = {
      inputValue: "",
      selectedItems: getSelectedItems(props),
      items: props.items || [],
    };
  }

  static getDerivedStateFromProps(props, state) {
    if (
      props.field &&
      props.field.value &&
      props.field.value.length !== state.selectedItems.length
    ) {
      return {
        inputValue: "",
        selectedItems: getSelectedItems(props),
      };
    }
    return null;
  }

  _input;
  _inputWrapper;

  /**
   * Used by onSearch to change the dropdown items that exist
   * @param {*} items
   */
  setItems = (items: Items): void => {
    if (!(items && Array.isArray(items))) {
      return;
    }
    this.setState({ items });
  };

  /**
   * Callback Fn from Downshift that exposes
   * state and event triggers such as keyboard
   * and mouse events
   */
  stateReducer = (state, changes) => {
    // this prevents the menu from being closed when the user
    // selects an item with a keyboard or mouse
    switch (changes.type) {
      case Downshift.stateChangeTypes.keyDownEnter:
      case Downshift.stateChangeTypes.clickItem:
        // Check the prop to decide whether we close the dropdown after an item is selected
        return {
          ...changes,
          isOpen: this.props.closeOnSelect ? false : state.isOpen,
        };
      case Downshift.stateChangeTypes.blurInput:
        // If a user tabs through the input, we want to make sure it's marked as
        // `touched` for formik's validation to run
        if (this.props.form && this.props.field) {
          this.props.form.setFieldTouched(this.props.field.name, true);
        }
        // Always close the dropdown on blur
        return {
          isOpen: false,
        };
      default:
        return changes;
    }
  };

  /**
   * Callback Fn from downshift that exposes changes
   */
  handleStateChange = (changes, downshiftStateAndHelpers) => {
    if (!downshiftStateAndHelpers.isOpen) {
      this.setState({ inputValue: "" });
    }
    if (changes.hasOwnProperty("selectedItem")) {
      this.onAddTag(changes.selectedItem);
    }
  };

  onAddTag = (item) => {
    const {
      field: { name, value },
      form: { setFieldValue },
      tags,
    } = this.props;
    const { inputValue, selectedItems } = this.state;

    if (item) {
      const updatedSelectedItems = [...selectedItems];
      updatedSelectedItems.push(item);
      this.setState({
        selectedItems: updatedSelectedItems,
        inputValue: "",
      });

      const updatedValue = [...value];
      updatedValue.push(item.value);
      setFieldValue(name, updatedValue);
    } else if (tags && inputValue !== "" && !value.includes(inputValue)) {
      const updatedSelectedItems = [...selectedItems];
      updatedSelectedItems.push({ label: inputValue, value: inputValue });
      this.setState({
        selectedItems: updatedSelectedItems,
        inputValue: "",
      });

      const updatedValue = [...value];
      updatedValue.push(inputValue);
      setFieldValue(name, updatedValue);
    }
  };

  onRemoveTag = (item) => {
    const {
      field: { name, value },
      form: { setFieldValue },
    } = this.props;
    const { selectedItems } = this.state;

    const stateIndex =
      selectedItems &&
      selectedItems.findIndex((val) => val.value === item.value);
    const updatedSelectedItems = [...selectedItems];
    updatedSelectedItems.splice(stateIndex, 1);
    this.setState({ selectedItems: updatedSelectedItems });

    const index = value && value.findIndex((val) => val === item.value);
    const updatedValue = [...value];
    updatedValue.splice(index, 1);
    setFieldValue(name, updatedValue);
  };

  popValue() {
    const {
      field: { name, value },
      form: { setFieldValue },
    } = this.props;
    const { selectedItems } = this.state;
    const updatedSelectedItems = [...selectedItems];
    updatedSelectedItems.pop();
    this.setState({ selectedItems: updatedSelectedItems });

    const updatedValue = [...value];
    updatedValue.pop();
    setFieldValue(name, updatedValue);
  }

  /**
   * Change state to reflect new inputValue typed by user.
   *
   * If onSearch is passed into props, call onSearch to allow AJAX autocomplete
   */
  onInputChange = (e) => {
    const inputValue = e.target.value;
    this.setState({ inputValue });

    if (this.props.onSearch) {
      this.props.onSearch({
        inputValue,
        originalItems: this.props.items,
        currentItems: this.state.items,
        setItems: this.setItems,
      });
    }
  };

  onInputKeyDown = (event) => {
    const currentValue = event.target.value;
    switch (event.keyCode) {
      case 8: // backspace
        if (!currentValue) {
          event.preventDefault();
          this.popValue();
        }
        return;
      case 9: // tab
        // Blur input, close dropdown
        return;
      case 13: // enter
        if (currentValue !== "") {
          event.preventDefault();
          // This is wrapped in a setTimeout to force it to trigger *after* the
          // built-in downshift state reducer fires (fixes a timing issue if a
          // user has typed-ahead *and* keyboard-selected an item from the dropdown)
          setTimeout(() => {
            this.onAddTag();
          }, 1);
        }
        return;
      default:
        return;
    }
  };

  /**
   * When user clicks on Dropdown, focus on the text input inside the dropdown
   */
  onWrapperClick = (e) => {
    e.stopPropagation();
    e.preventDefault();
    this.focusInput();
  };

  focusInput() {
    this._input.focus();
    if (typeof this._input.getInput === "function") {
      this._input.getInput().focus();
    }
  }

  inputRef = (c) => {
    this._input = c;
  };

  inputWrapperRef = (c) => {
    this._inputWrapper = c;
  };

  renderArrowIcon(isOpen) {
    const {
      classes,
      theme: { colors },
    } = this.props;

    return (
      <svg
        className={classes.arrow}
        width={24}
        height={24}
        viewBox="0 0 1792 1792"
        transform={isOpen ? "rotate(180)" : null}
      >
        <path
          fill={colors.gray300 || "currentColor"}
          d="M1395 736q0 13-10 23l-466 466q-10 10-23 10t-23-10l-466-466q-10-10-10-23t10-23l50-50q10-10 23-10t23 10l393 393 393-393q10-10 23-10t23 10l50 50q10 10 10 23z"
        />
      </svg>
    );
  }

  render() {
    const {
      classes,
      className,
      field: { name, value },
      form: { errors, touched },
      label,
      withScaffold = true,
    } = this.props;

    const { selectedItems, items } = this.state;

    return (
      <Downshift
        itemToString={itemToString}
        stateReducer={this.stateReducer}
        onStateChange={this.handleStateChange}
        selectedItem={value}
      >
        {({
          getInputProps,
          getItemProps,
          isOpen,
          highlightedIndex,
          openMenu,
          closeMenu,
        }) => {
          const Input = (
            <div
              ref={this.inputWrapperRef}
              className={classNames(classes.root, className)}
              onClick={this.onWrapperClick}
            >
              <div className={classes.tagWrapper}>
                {selectedItems.map((item) => {
                  return (
                    <Fragment key={item.value}>
                      <Tag tag={item} onRemove={this.onRemoveTag} />
                      <input type="hidden" name={name} value={item.value} />
                    </Fragment>
                  );
                })}
                <AutosizeInput
                  {...getInputProps({
                    ref: this.inputRef,
                    inputClassName: classes.input,
                    onChange: this.onInputChange,
                    onFocus: !isOpen ? openMenu : closeMenu,
                    onKeyDown: this.onInputKeyDown,
                    value: this.state.inputValue,
                  })}
                />
              </div>
              {this.renderArrowIcon(isOpen)}
            </div>
          );

          const Dropdown = (
            <div className={classes.dropdown}>
              {items
                .filter(
                  (i) =>
                    this.state.selectedItems &&
                    this.state.selectedItems.findIndex(
                      (val) => val.value === i.value
                    ) === -1 &&
                    (!this.state.inputValue ||
                      i.label
                        .toLowerCase()
                        .includes(this.state.inputValue.toLowerCase()))
                )
                .map((item, index) => (
                  <div
                    key={item.value}
                    className={classNames({
                      [classes.dropdownItem]: true,
                      [classes.dropdownItemSelected]:
                        value &&
                        value.findIndex((val) => val.value === item.value) > -1,
                      [classes.dropdownItemActive]: highlightedIndex === index,
                    })}
                    {...getItemProps({
                      index,
                      item,
                    })}
                  >
                    {itemToString(item)}
                  </div>
                ))}
            </div>
          );

          return withScaffold ? (
            <div>
              <Scaffold
                label={label}
                error={get(touched, name) && get(errors, name)}
              >
                {Input}
                {isOpen && Dropdown}
              </Scaffold>
            </div>
          ) : (
            <div>
              {Input}
              {isOpen && Dropdown}
            </div>
          );
        }}
      </Downshift>
    );
  }
}

const styles = (theme) => ({
  root: {
    display: "flex",
    alignItems: "center",
    width: "100%",
    minHeight: "40px",
    backgroundColor: theme.colors.white,
    border: `1px solid ${theme.colors.gray300}`,
    padding: "2px 12px 2px 8px",
    cursor: (props) => (props.disabled ? "not-allowed" : "default"),
  },
  tagWrapper: {
    display: "flex",
    alignItems: "center",
    overflowX: "auto",
  },
  input: {
    color: "inherit",
    fontFamily: "inherit",
    fontWeight: "400",
    fontSize: "16px",
    lineHeight: "20px",
    border: "none",
    outline: "none",
    cursor: "inherit",
    boxShadow: "none", // Reset default inputs for mozilla
    "-webkit-appearance": "none", // Reset default browser styles
    "-moz-appearance": "none", // Reset default browser styles
  },
  arrow: {
    flexShrink: "0",
    display: "block",
    marginLeft: "auto",
  },
  dropdown: {
    position: "absolute",
    zIndex: theme.zIndex["selectDropdown"],
    width: "100%",
    maxHeight: "200px",
    cursor: "default",
    backgroundColor: theme.colors.white,
    border: `1px solid ${theme.colors.gray300}`,
    boxShadow: theme.globalBoxShadow, // Add back focus style
    marginTop: "-1px", // force overlapping border with input
    overflowY: "scroll",
    WebkitOverflowScrolling: "touch",
  },
  dropdownItem: {
    fontSize: "16px",
    lineHeight: "20px",
    fontWeight: "400",
    padding: "8px 12px",
  },
  dropdownItemActive: {
    backgroundColor: theme.colors.gray100,
  },
  dropdownItemSelected: {
    backgroundColor: theme.colors.gray200,
  },
});

export default withStyles(styles, { name: "EuiMultiSelect" })(MultiSelect);
