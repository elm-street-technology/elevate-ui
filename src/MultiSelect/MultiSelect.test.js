import React from "react";
import renderer from "react-test-renderer";

import ThemeProvider from "../ThemeProvider";
import MultiSelect from "./";

const roygbiv = [
  { label: "Red", value: "red" },
  { label: "Orange", value: "orange" },
  { label: "Yellow", value: "yellow" },
  { label: "Green", value: "green" },
  { label: "Blue", value: "blue" },
  { label: "Indigo", value: "indigo" },
  { label: "Violet", value: "violet" },
];

const WrappedMultiSelect = (props) => (
  <ThemeProvider>
    <MultiSelect
      field={{
        name: "",
        onChange: () => {},
        onBlur: () => {},
        value: [],
      }}
      form={{
        errors: [],
        setFieldValue: () => {},
        setFieldTouched: () => {},
        touched: false,
      }}
      {...props}
    />
  </ThemeProvider>
);

test("renders without crashing, matches the snapshot", () => {
  const component = renderer.create(
    <WrappedMultiSelect id="color" items={roygbiv} />
  );
  const componentJSON = component.toJSON();
  expect(componentJSON).toMatchSnapshot();
});
