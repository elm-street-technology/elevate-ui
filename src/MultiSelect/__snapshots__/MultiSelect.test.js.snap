// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing, matches the snapshot 1`] = `
<div>
  <div
    className="Scaffold-root-0-1-10"
  >
    <div
      className="MultiSelect-wrapper-0-1-1 MultiSelect-wrapper-0-1-9"
      onClick={[Function]}
    >
      <div
        className="MultiSelect-tagWrapper-0-1-2"
      >
        <div
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-activedescendant={null}
            aria-autocomplete="list"
            aria-expanded={false}
            autoComplete="off"
            className="MultiSelect-root-0-1-3"
            id="downshift-0-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            role="combobox"
            style={
              Object {
                "boxSizing": "content-box",
                "width": "1px",
              }
            }
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          >
            
          </div>
        </div>
      </div>
      <svg
        className="MultiSelect-arrow-0-1-4"
        height={24}
        transform={null}
        viewBox="0 0 1792 1792"
        width={24}
      >
        <path
          d="M1395 736q0 13-10 23l-466 466q-10 10-23 10t-23-10l-466-466q-10-10-10-23t10-23l50-50q10-10 23-10t23 10l393 393 393-393q10-10 23-10t23 10l50 50q10 10 10 23z"
          fill="#ccd2d8"
        />
      </svg>
    </div>
  </div>
</div>
`;
